File: edu/yu/compilers/intermediate/symtable/SymTableEntry.java
/**
 * <h1>SymTableEntryImpl</h1>
 * <p>An implementation of a symbol table entry.</p>
 * <p>Adapted from:</p>
 * <p>Copyright (c) 2020 by Ronald Mak</p>
 * <p>For instructional purposes only.  No warranties.</p>
 */

package edu.yu.compilers.intermediate.symtable;

import edu.yu.compilers.intermediate.type.Typespec;

import java.util.ArrayList;


public class SymTableEntry {
    private final String name;                     // entry name
    private final SymTable symTable;                   // parent symbol table
    private final ArrayList<Integer> lineNumbers;  // source line numbers
    private Kind kind;                       // what kind of identifier
    private Typespec typespec;               // type specification
    private int slotNumber;                  // local variables array slot number
    private EntryInfo info;                  // entry information

    /**
     * Constructor.
     *
     * @param name     the name of the entry.
     * @param kind     the kind of entry.
     * @param symTable the symbol table that contains this entry.
     */
    public SymTableEntry(String name, Kind kind, SymTable symTable) {
        this.name = name;
        this.kind = kind;
        this.symTable = symTable;
        this.lineNumbers = new ArrayList<>();

        // Initialize the appropriate entry information.
        switch (kind) {
            case CONSTANT, ENUMERATION_CONSTANT, VARIABLE, RECORD_FIELD, VALUE_PARAMETER -> info = new ValueInfo();
            case PROGRAM, PROCEDURE, FUNCTION -> {
                info = new RoutineInfo();
                ((RoutineInfo) info).parameters = new ArrayList<>();
                ((RoutineInfo) info).subroutines = new ArrayList<>();
            }
            default -> {
            }
        }
    }

    /**
     * Get the name of the entry.
     *
     * @return the name.
     */
    public String getName() {
        return name;
    }

    /**
     * Get the kind of entry.
     *
     * @return the kind.
     */
    public Kind getKind() {
        return kind;
    }

    /**
     * Set the kind of entry.
     *
     * @param kind the kind to set.
     */
    public void setKind(Kind kind) {
        this.kind = kind;
    }

    /**
     * Get the symbol table that contains this entry.
     *
     * @return the symbol table.
     */
    public SymTable getSymTable() {
        return symTable;
    }

    /**
     * Get the slot number of the local variables array.
     *
     * @return the number.
     */
    public int getSlotNumber() {
        return slotNumber;
    }

    /**
     * Set the slot number of the local variables array.
     *
     * @param slotNumber the number to set.
     */
    public void setSlotNumber(int slotNumber) {
        this.slotNumber = slotNumber;
    }

    /**
     * Get the type specification of the entry.
     *
     * @return the type specification.
     */
    public Typespec getType() {
        return typespec;
    }

    /**
     * Set the type specification.
     *
     * @param typespec the type specification to set.
     */
    public void setType(Typespec typespec) {
        this.typespec = typespec;
    }

    /**
     * Get the arraylist of source line numbers for the entry.
     *
     * @return the arraylist.
     */
    public ArrayList<Integer> getLineNumbers() {
        return lineNumbers;
    }

    /**
     * Append a source line number to the entry.
     *
     * @param lineNumber the line number to append.
     */
    public void appendLineNumber(int lineNumber) {
        lineNumbers.add(lineNumber);
    }

    /**
     * Get the data value stored with this entry.
     *
     * @return the data value.
     */
    public Object getValue() {
        return ((ValueInfo) info).value;
    }

    /**
     * Set the data value into this entry.
     *
     * @param value the value to set.
     */
    public void setValue(Object value) {
        ((ValueInfo) info).value = value;
    }

    /**
     * Get the routine code.
     *
     * @return the code.
     */
    public Routine getRoutineCode() {
        return ((RoutineInfo) info).code;
    }

    /**
     * Set the routine code.
     *
     * @param code the code to set.
     */
    public void setRoutineCode(Routine code) {
        ((RoutineInfo) info).code = code;
    }

    /**
     * Get the routine's symbol table.
     *
     * @return the symbol table.
     */
    public SymTable getRoutineSymTable() {
        return ((RoutineInfo) info).symTable;
    }

    /**
     * Set the routine's symbol table.
     *
     * @param symTable the symbol table to set.
     */
    public void setRoutineSymTable(SymTable symTable) {
        ((RoutineInfo) info).symTable = symTable;
    }

    /**
     * Get the arraylist of symbol table entries of the routine's formal parameters.
     *
     * @return the arraylist.
     */
    public ArrayList<SymTableEntry> getRoutineParameters() {
        return ((RoutineInfo) info).parameters;
    }

    /**
     * Set the arraylist symbol table entries of parameters of the routine.
     *
     * @param parameters the arraylist to set.
     */
    public void setRoutineParameters(ArrayList<SymTableEntry> parameters) {
        ((RoutineInfo) info).parameters = parameters;
    }

    /**
     * Get the arraylist of symbol table entries of the nested subroutines.
     *
     * @return the arraylist.
     */
    public ArrayList<SymTableEntry> getSubroutines() {
        return ((RoutineInfo) info).subroutines;
    }

    /**
     * Append to the arraylist of symbol table entries of the nested subroutines.
     *
     * @param subroutineId the symbol table entry of the subroutine to append.
     */
    public void appendSubroutine(SymTableEntry subroutineId) {
        ((RoutineInfo) info).subroutines.add(subroutineId);
    }

    /**
     * Get the routine's executable code.
     *
     * @return the executable code.
     */
    public Object getExecutable() {
        return ((RoutineInfo) info).executable;
    }

    /**
     * Set the routine's executable code.
     *
     * @param executable the executable code to set.
     */
    public void setExecutable(Object executable) {
        ((RoutineInfo) info).executable = executable;
    }

    /**
     * What kind of identifier.
     */
    public enum Kind {
        CONSTANT, ENUMERATION_CONSTANT, TYPE, VARIABLE, RECORD_FIELD, VALUE_PARAMETER, REFERENCE_PARAMETER, PROGRAM_PARAMETER, PROGRAM, PROCEDURE, FUNCTION, UNDEFINED;

        public String toString() {
            return super.toString().toLowerCase();
        }
    }

    /**
     * Which routine.
     */
    public enum Routine {
        DECLARED, FORWARD, READ, READLN, WRITE, WRITELN, ABS, ARCTAN, CHR, COS, EOF, EOLN, EXP, LN, ODD, ORD, PRED, ROUND, SIN, SQR, SQRT, SUCC, TRUNC,
    }

    /**
     * Entry information interface.
     */
    private interface EntryInfo {
    }

    /**
     * Value information.
     */
    private static class ValueInfo implements EntryInfo {
        private Object value;
    }

    /**
     * Routine information.
     */
    private static class RoutineInfo implements EntryInfo {
        private Routine code;                          // routine code
        private SymTable symTable;                     // routine's symbol table
        private ArrayList<SymTableEntry> parameters;   // routine's formal parameters
        private ArrayList<SymTableEntry> subroutines;  // symTable entries of subroutines
        private Object executable;                     // routine's executable code
    }
}

File: edu/yu/compilers/intermediate/symtable/SymTableEntry.py
import abc
from edu.yu.compilers.intermediate.symtable.Kind import Kind
from edu.yu.compilers.intermediate.symtable.Routine import Routine
from enum import Enum
from edu.yu.compilers.intermediate.type.Typespec import Typespec



class SymTableEntry:
    # Entry information interface.
    class EntryInfo(abc.ABC):
        dud = None


    # Value information.
    class ValueInfo(EntryInfo):
        value = None

    # Routine information.
    class RoutineInfo(EntryInfo):
        code = None  # routine code
        symtable = None  # routine's symbol table
        parameters = None  # routine's formal parameters
        subroutines = None  # symTable entries of subroutines
        executable = None  # routine's executable code
        inline = False

    # Constructor.
      #
      # @param name     the name of the entry.
      # @param kind     the kind of entry.
      # @param symTable the symbol table that contains this entry.

    def __init__(self, name, kind, symTable) :
        self.name = name
        self.kind = kind
        self.symTable = symTable
        self.lineNumbers = []
        self.typeSpec = None
        self.slotNumber = None

        # Initialize the appropriate entry information.
        if self.kind == Kind.CONSTANT or self.kind == Kind.ENUMERATION_CONSTANT or self.kind == Kind.VARIABLE or self.kind == Kind.RECORD_FIELD or self.kind == Kind.VALUE_PARAMETER :
            self.info = self.ValueInfo()
        elif self.kind == Kind.PROGRAM or self.kind == Kind.PROCEDURE or self.kind == Kind.FUNCTION:
            self.info = self.RoutineInfo()
            self.info.inline = False
            self.info.parameters = []
            self.info.subroutines = []

    # Get the name of the entry.
    # @return the name.
    def getName(self):
        return self.name


    # Get the kind of entry.
    #
    # @return the 

    def getKind(self) :
        return self.kind

    # Set the kind of entry.
    #
    # @param kind the kind to set.

    def setKind(self, kind) :
        self.kind = kind


    # Get the symbol table that contains this entry.
    #
    # @return the symbol table.

    def getSymTable(self) :
        return self.symTable


    # Get the slot number of the local variables array.
    #
    # @return the number.

    def getSlotNumber(self) :
        return self.slotNumber


    # Set the slot number of the local variables array.
    #
    # @param slotNumber the number to set.

    def setSlotNumber(self, slotNumber) :
        self.slotNumber = slotNumber


    # Get the type specification of the entry.
    #
    # @return the type specification.

    def getType(self) :
        return self.typespec


    # Set the type specification.
    #
    # @param typespec the type specification to set.

    def setType(self, typespec) :
        self.typespec = typespec



    # Get the arraylist of source line numbers for the entry.
    #
    # @return the arraylist.

    def getLineNumbers(self) :
        return self.lineNumbers


    # Append a source line number to the entry.
    #
    # @param lineNumber the line number to append.

    def appendLineNumber(self, lineNumber) :
        self.lineNumbers.append(lineNumber)


    # Get the data value stored with this entry.
    #
    # @return the data value.

    def getValue(self) :
        return self.info.value


    # Set the data value into this entry.
    #
    # @param value the value to set.

    def setValue(self, value) :
        self.info.value = value


    # Get the routine code.
    #
    # @return the code.

    def getRoutineCode(self) :
        return self.info.code


    # Set the routine code.
    #
    # @param code the code to set.

    def setRoutineCode(self, code) :
        self.info.code = code


    # Get the routine's symbol table.
    #
    # @return the symbol table.

    def getRoutineSymTable(self) :
        return self.info.symTable


     # Set the routine's symbol table.
     #
     # @param symTable the symbol table to set.

    def setRoutineSymTable(self, symTable) :
        self.info.symTable = symTable

    # Get the arraylist of symbol table entries of the routine's formal parameters.
    #
    # @return the arraylist.

    def getRoutineParameters(self) :
        return self.info.parameters

    # Set the arraylist symbol table entries of parameters of the routine.
    #
    # @param parameters the arraylist to set.
    def setRoutineParameters(self, parameters) :
        self.info.parameters = parameters

    # Get the arraylist of symbol table entries of the nested subroutines.
    #
    # @return the arraylist.

    def getSubroutines(self) :
        return self.info.subroutines

    # Append to the arraylist of symbol table entries of the nested subroutines.
    #
    # @param subroutineId the symbol table entry of the subroutine to append.

    def appendSubroutine(self, subroutineId) :
        self.info.subroutines.append(subroutineId)

    # Get the routine's executable code.
    #
    # @return the executable code.
    def getExecutable(self) :
        return self.info.executable

    # Set the routine's executable code.
    #
    # @param executable the executable code to set.
    def setExecutable(self, executable) :
        self.info.executable = executable

    def isInline(self):
        return self.info.inline;

    def setInline(self, inline):
        self.info.inline = inline

